beetsplug/fish.py

```diff
@@ -13,329 +14,393 @@
-                cmd_need_arg = " -r " if option.nargs in [1] else ""
-                cmd_helpstr = (
-                    (" -d " + wrap(" ".join(option.help.split())))
-                    if option.help
-                    else ""
-                )
-                cmd_arglist = (
-                    (" -a " + wrap(" ".join(option.choices)))
-                    if option.choices
-                    else ""
+        if extra_comp_fields:
+            # The set of values for every user-specified extra field.
+            extra_values: Dict[str, set[str]] = dict.fromkeys(
+                extra_comp_fields, set()
+            )
+            for item in lib.items():
+                for key, val in extra_values.items():
+                    val.add(str(item[key]))
+
+            for field in sorted(extra_values.keys()):
+                values = extra_values[field]
+                field_text = fish_escape(field, style="var")
+                value_array = f"__fish_beet_field_{field_text}"
+                # The field name is explicit since we are adding to the token the user
+                # is actively writing to -- usually we only examine the preceding ones.
+                script.set_array(
+                    value_array, sorted(f"{field}:{v}" for v in values)
                 )
-
-                word += (
-                    " ".join(
-                        BL_USE3.format(
-                            name,
-                            (
-                                cmd_need_arg
-                                + cmd_s
-                                + cmd_l
-                                + " -f "
-                                + cmd_arglist
-                            ),
-                            cmd_helpstr,
-                        ).split()
-                    )
-                    + "\n"
+                in_command = "__fish_beet_subcommand"
+                in_value = f"__fish_beet_metadata_param {fish_escape(field)}"
+                script.complete(
+                    values=f"${value_array}",
+                    conditions=[in_command, in_value],
+                    description="known metadata value",
                 )
 
-            word = word + " ".join(
-                BL_USE3.format(
-                    name,
-                    ("-s " + "h " + "-l " + "help" + " -f "),
-                    ("-d " + wrap("print help") + "\n"),
-                ).split()
-            )
-    return word
-
-
-def clean_whitespace(word):
-    # Remove excess whitespace and tabs in a string
-    return " ".join(word.split())
-
-
-def wrap(word):
-    # Need " or ' around strings but watch out if they're in the string
-    sptoken = '"'
-    if ('"') in word and ("'") in word:
-        word.replace('"', sptoken)
-        return '"' + word + '"'
-
-    tok = '"' if "'" in word else "'"
-    return tok + word + tok
+        # Write the buffered text to the file.
+        output.write_text(script.getvalue())
```

Comment:
As a user I would probably expect the completions to be printed to stdout, since that's what most other tools do. For example, some users may be using a different path to store their completions.

Reply:
It's possible to configure a different path using `-o`.  As a Fish user, I find the current behaviour quite convenient, since I'm not fully familiar with where these completions need to go.  I can add support for printing to `stdout` using `-o -`, but I still think the current behaviour is a good default.  Also note backward compatibility -- users may rely on the current behaviour as part of e.g. periodic scripts.